app/scripts/metamask-controller.js

```diff
@@ -4431,43 +4431,49 @@ export default class MetamaskController extends EventEmitter {
   }
 
   async _addAccountsWithBalance() {
-    // Scan accounts until we find an empty one
-    const chainId = getCurrentChainId({
-      metamask: this.networkController.state,
-    });
-    const ethQuery = new EthQuery(this.provider);
-    const accounts = await this.keyringController.getAccounts();
-    let address = accounts[accounts.length - 1];
-
-    for (let count = accounts.length; ; count++) {
-      const balance = await this.getBalance(address, ethQuery);
-
-      if (balance === '0x0') {
-        // This account has no balance, so check for tokens
-        await this.tokenDetectionController.detectTokens({
-          chainIds: [chainId],
-          selectedAddress: address,
-        });
+    try {
+      // Scan accounts until we find an empty one
+      const chainId = getCurrentChainId({
+        metamask: this.networkController.state,
+      });
+      const ethQuery = new EthQuery(this.provider);
+      const accounts = await this.keyringController.getAccounts();
+      let address = accounts[accounts.length - 1];
+
+      for (let count = accounts.length; ; count++) {
+        const balance = await this.getBalance(address, ethQuery);
+
+        if (balance === '0x0') {
+          // This account has no balance, so check for tokens
+          await this.tokenDetectionController.detectTokens({
+            chainIds: [chainId],
+            selectedAddress: address,
+          });
+
+          const tokens =
+            this.tokensController.state.allTokens?.[chainId]?.[address];
+          const detectedTokens =
+            this.tokensController.state.allDetectedTokens?.[chainId]?.[address];
 
-        const tokens =
-          this.tokensController.state.allTokens?.[chainId]?.[address];
-        const detectedTokens =
-          this.tokensController.state.allDetectedTokens?.[chainId]?.[address];
-
-        if (
-          (tokens?.length ?? 0) === 0 &&
-          (detectedTokens?.length ?? 0) === 0
-        ) {
-          // This account has no balance or tokens
-          if (count !== 1) {
-            await this.removeAccount(address);
+          if (
+            (tokens?.length ?? 0) === 0 &&
+            (detectedTokens?.length ?? 0) === 0
+          ) {
+            // This account has no balance or tokens
+            if (count !== 1) {
+              await this.removeAccount(address);
+            }
+            break;
           }
-          break;
         }
-      }
 
-      // This account has assets, so check the next one
-      address = await this.keyringController.addNewAccount(count);
+        // This account has assets, so check the next one
+        address = await this.keyringController.addNewAccount(count);
+      }
+    } catch (e) {
+      log.warn(`Failed to add accounts with balance. Error: ${e}`);
+    } finally {
+      this.preferencesController.setHasFinishedAddingAccountsWithBalance(true);
```

Comment:
In order to prevent conflicts with different actors adding accounts at the end of onboarding, we're only dispatching account syncing when `_addAccountsWithBalance` is done.